app/src/org/commcare/activities/DispatchActivity.java

```diff
@@ -140,34 +140,33 @@ private void dispatch() {
         } else {
             // Note that the order in which these conditions are checked matters!!
             try {
-                ApplicationRecord currentRecord = currentApp.getAppRecord();
-                if (currentApp.getAppResourceState() == CommCareApplication.STATE_CORRUPTED) {
-                    // The seated app is damaged or corrupted
-                    handleDamagedApp();
-                } else if (!currentRecord.isUsable()) {
-                    // The seated app is unusable (means either it is archived or is
-                    // missing its MM or both)
-                    boolean unseated = handleUnusableApp(currentRecord);
-                    if (unseated) {
-                        // Recurse in order to make the correct decision based on the new state
-                        dispatch();
-                    }
-                } else if (!CommCareApplication._().getSession().isActive()) {
-                    // The user is not logged in
-                    launchLoginScreen();
-                } else if (this.getIntent().hasExtra(SESSION_REQUEST)) {
-                    // CommCare was launched from an external app, with a session descriptor
-                    handleExternalLaunch();
-                } else if (this.getIntent().hasExtra(AndroidShortcuts.EXTRA_KEY_SHORTCUT) &&
-                        !shortcutExtraWasConsumed) {
-                    // CommCare was launched from a shortcut
-                    handleShortcutLaunch();
-                } else {
-                    launchHomeScreen();
-                }
+                CommCareApplication._().getSession();
             } catch (SessionUnavailableException sue) {
                 launchLoginScreen();
             }
+
+            ApplicationRecord currentRecord = currentApp.getAppRecord();
+            if (currentApp.getAppResourceState() == CommCareApplication.STATE_CORRUPTED) {
```

Comment:
Hm, so when I originally wrote this, I intentionally ordered it to check for these 2 states (if the app resource state is corrupted, or the currently seated app is not usable), _before_ checking for session expiration, so that you couldn't end up on the login screen with a corrupted or unusable app as the seated app. I don't have hard evidence that this is necessary, but I think we probably want to still preserve that behavior.